### DISCUSSION TITLE
RFC: web worker API / implementation

### DISCUSSION THREAD

**Discussion by ntoll** [pyscript/pyscript#1113] [Proposals]
We should be able to specify that the code relating to a certain `<py-script>` tag is to be run in a non-blocking web-worker. Both @pauleveritt (with Pyodide) and @ntoll / @marimeireles (with MicroPython) have extensively explored how this might happen. Now we've some working experiments, we should consolidate the lessons learned into a unified, simple and clearly defined API/implementation.

This discussion has been started to allow us to pool resources, share and refine suggestions and come to a consensus as to the way forward.

Please add suggestions, example code, pointers to experiments and so on, below. 🚀

**Comment by pauleveritt** [comment]
A fun point to add.

If you're doing a worker, you might want ESM support, because doing everything in one module is kind of lame. However, Firefox [is a mere 7 years behind](https://bugzilla.mozilla.org/show_bug.cgi?id=1247687) on implementing it. Seems it got partially done, then backed out.

This can be solved using a bundler which flattens into one file. But that triggers some other concerns.

So, just an FYI on an edge case where I flailed for a day and punted.

**Reply by ntoll** [reply]
Ack. Good to know. Tech is all about compromises... and this process is all about discovering what they might be. Thanks for the heads up. :rocket:

**Comment by pauleveritt** [comment]
Another note...if you're doing any non-E2E testing, workers introduce a wrench. I covered it a bit in my writeup. I used a different stack than PyScript but some of the concepts are the same.

**Reply by ntoll** [reply]
It'd be really helpful to add a link to your write-up so everyone has visibility. :+1:

**Comment by pauleveritt** [comment]
_Oops, sorry_. I did a "diary-driven development"[ experiment on "Pyodide Components"](https://pyodide-components.readthedocs.io/en/latest/project_init.html). It touches on a number of points: workers, testing, writing components in Python, returning HTML string to Idiomorph (instead of DOM-like), etc.

I have a few more sections written but not published: injection, server-side pre-rendering, etc.
